paymentapi/src/main/java/fi/hel/verkkokauppa/payment/paytrail/validation/PaytrailPaymentReturnValidator.java

```diff
@@ -17,15 +24,26 @@
 @Slf4j
 public class PaytrailPaymentReturnValidator {
 
-    @Value("${paytrail.merchant.secret:}")
-    private String secretKey;
+    @Value("${paytrail.aggregate.merchant.secret:}")
+    private String aggregateSecretKey;
 
+    @Autowired
+    private PaymentRepository paymentRepository;
 
-    public boolean validateChecksum(Map<String,String> checkoutParams, String signature, String orderNumber) {
-        if (buildSignatureFromReturnData(checkoutParams).equals(signature)) {
+    @Autowired
+    private CommonServiceConfigurationClient commonServiceConfigurationClient;
+
+
+    public boolean validateChecksum(Map<String,String> checkoutParams, String merchantId, String signature, String paymentId) {
+        Payment payment = paymentRepository.findById(paymentId).orElseThrow(() -> new CommonApiException(
+                HttpStatus.NOT_FOUND,
+                new Error("payment-not-found", "payment with id [" + paymentId + "] not found")
+        ));
+        String createdSignature = buildSignatureFromReturnData(checkoutParams, merchantId, payment.isShopInShopPayment());
+        if (createdSignature.equals(signature)) {
             return true;
         } else {
-            log.debug("paytrail payment validation failed, orderNumber: " + orderNumber);
```

Comment:
orderNumber = paymentId. This is referred to as **paymentId** more in the whole codebase, so thats why I changed it.

Reply:
we need the paymentId because it has the timestamp at end of it?

Reply:
Yeh much better then we know which payment was the wrong type etc.